### PR TITLE
amalgamate checking of attributes and extra types

### PR DESCRIPTION
**Pull request by omsmith**

### CODE REVIEW

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs [hunk 1]
```diff
@@ -165,6 +160,29 @@ internal sealed class ImmutabilityContext {
 			}
 		}
 
+		private bool TryGetImmutableTypeInfo(
```
Comment (omsmith) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs:163]: Slight difference, in that `m_extraImmutableTypes` isn't considered if a type already has `[Immutable*]` attributes

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs [hunk 2]
```diff
@@ -165,6 +160,29 @@ internal sealed class ImmutabilityContext {
...
+			if( fromAttributes != ImmutableTypeKind.None ) {
+				info = new ImmutableTypeInfo(
+					kind: fromAttributes,
+					type: type
+				);
+				return true;
+			}
+
+			// Check if we were otherwise told that this type is immutable
+			if( m_extraImmutableTypes.TryGetValue( type, out info ) ) {
```
Comment (omsmith) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs:178]: Should probably eventually look at `OriginalDefinition` as well, but will save making such adjustments for later

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs [hunk 3]
```diff
@@ -165,6 +160,29 @@ internal sealed class ImmutabilityContext {
...
+				return true;
+			}
+
+			info = default;
+			return false;
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs:183]:
```suggestion
			return m_extraImmutableTypes.TryGetValue( type, out info );
```

I think?

**Review by j3parker — DISMISSED** [review]
Nice

**Comment by j3parker** [comment]
@mthjones this is to support immutable container types -- we want to ask "how" a type is immutable and then check if that's sufficiently immutable for us. It shouldn't matter if its done with attributes or done via `m_extraImmutableTypes`; `ImmutableTypeInfo` abstracts this away.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs [hunk 4]
```diff
@@ -84,15 +84,10 @@ internal sealed class ImmutabilityContext {
...
-
-			// Check if we were otherwise told that this type is (sufficiently)
-			// immutable.
-			if ( m_extraImmutableTypes.TryGetValue( type, out var info ) && info.Kind.HasFlag( kind ) ) {
-				return true;
+			if( TryGetImmutableTypeInfo( type, out ImmutableTypeInfo info ) ) {
+				if( info.Kind.HasFlag( kind ) ) {
+					return true;
+				}
 			}
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs:91]: @mthjones this is to support immutable container types better -- we want to ask "how" a type is immutable and then check if that's sufficiently immutable for us. It shouldn't matter if its done with attributes or done via m_extraImmutableTypes; ImmutableTypeInfo abstracts this away.

**Review by mthjones — DISMISSED** [review]

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs (hunk 3 above)
Comment (omsmith) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs:183]: Yeah, just depends on your feelings about breakpoints

**Review by j3parker — APPROVED** [review]

**Pull request merged**